### ISSUE TITLE
Feature request: Support network policies

### ISSUE THREAD

**Issue by olljanat** [acorn-io/acorn#456]
As Acorn can be used to deploy multiple containers as bundle it it would be nice to have also option to generate network policy which isolates application from others.

**Comment by ibuildthecloud** [comment]
Yes, this absolutely will come.  The design of acorn is such that all network communication paths are known and have some logical boundary.  We will either be generating network policies and/or service mesh config (istio/linkerd) to enforce this.  This is key future feature for the security posture of Acorn.  One of the great things about Acorn is we know so much about the application behavior that we can create good implicit rules.

**Comment by ibuildthecloud** [comment]
Researching this a bit, network policies would be extremely simple.  LInkerd looks straight forward too and Istio is rocket science like normal.  So we will start with NetworkPolicies and go from there.

**Comment by olljanat** [comment]
In case of one app with multiple containers it should be very simple yes. In case of multiple apps you will need also figure out some nice way to refer another app in Acornfile bur perhaps even that can be just name reference.

**Comment by g-linville** [comment]
## Design Document

### Requirements

- Traffic between different projects is blocked by default
- Traffic to published ports is allowed from all sources

### UX and API changes

- New flag for `acorn install` that disables the creation of NetworkPolicies (something like `--disable-network-policies`).
  - This might also result in a slight API change for the struct that handles these installation settings. I haven't looked at this part of the code yet.

### Visualization

![image](https://user-images.githubusercontent.com/53102776/222276250-c4c567c5-5481-48c5-ae44-07c107ad5256.png)

### Additional considerations

- Backwards compatibility: if we turn on NetworkPolicies by default (and I think we should), this has the potential to break some users' existing configurations, if they were relying on applications in separate projects to be able to talk to each other on non-published ports. We would definitely need to warn users about this.
- Tests: I should be able to write tests for the changes I have already made in this draft PR. I haven't written tests for a new handler yet but I'll reference the other ones and I imagine it will be relatively straightforward. I can also add an integration test to truly deploy two apps in separate projects and try to have one ping or curl the other, if that is something we think is worthwhile.
- Time estimate: I think this should be a fairly easy feature to add that we can easily land in the next release.

**Comment by cjellick** [comment]
> New flag for acorn install that disables the creation of NetworkPolicies (something like --disable-network-policies).

I am wondering if we shou;d we have it disabled by default and you need to explicitly enable it?

> Traffic between different projects is blocked by default

To nitpick this, I think it is more accurate to say that all ingress traffic to a project should be blocked. All egress allowed

**Comment by cjellick** [comment]
Publishing a port puts it on an ingress. thatll def make it publicly accessible. That makes sense to me.

Should *exposing* a port poke a hole for it in the network policy so that it can be reached by other workloads in the cluster?

**Comment by g-linville** [comment]
> > Traffic between different projects is blocked by default
> 
> To nitpick this, I think it is more accurate to say that all ingress traffic to a project should be blocked. All egress allowed

Yeah that is definitely a better way of phrasing it.

> Should exposing a port poke a hole for it in the network policy so that it can be reached by other workloads in the cluster?

Yeah I'm not sure about this. My current implementation does poke this hole, but like I suggested, it might be better to have the user provide the namespace where the ingress controller lives, and allow access only from that namespace (and maybe also a label selector that selects only the ingress controller pods).

**Comment by cjellick** [comment]
Not sure user specifying ingress namespace will suffice, particularly with the `expose` keyword (which doesnt involve ingress at all) and is explicitly defined as making a workload to other services in the cluster: https://docs.acorn.io/running/networking#expose-individual-ports

**Comment by g-linville** [comment]
This is what we decided on (for the time being): published and exposed ports will both have NetPols created for them that allow all ingress traffic.

**Comment by cjellick** [comment]
moving this into the v0.7.0 milestone

**Comment by cjellick** [comment]
moving this into the v0.7.0 milestone

**Comment by sangee2004** [comment]
Tested with acorn `version v0.6.0-21-gb1b07672+b1b07672`

On Acorn install , Network policy is enabled by default - `disableNetworkPolicies: false`
Project isolation for ping, http and ssh access works as expected . 
project1 - App1 - containers - web and ubuntu
project2 - App2 - containers - web and ubuntu
                  TestContainerApp - containers - testcon

From testcon 
1. Able to curl App1.web's pod ip
2. Able to ssh App1.web's ubuntu ip
3. Able to ping App1.web's pod ip and App1.web's ubuntu ip

4. Not able to curl App1.web's pod ip 
```
root@myapp:/app# curl 10.42.0.91    
curl: (7) Failed to connect to 10.42.0.91 port 80: Connection refused
```
6. Not able to ssh App1.web's ubuntu ip
```
root@myapp:/app# ssh root@10.42.0.90
ssh: connect to host 10.42.0.90 port 22: Connection refused
```
8. Not able to ping App1.web's pod ip and App1.web's ubuntu ip 
```PING 10.42.0.90 (10.42.0.90) 56(84) bytes of data.
From 10.42.0.1 icmp_seq=1 Destination Port Unreachable
From 10.42.0.1 icmp_seq=2 Destination Port Unreachable
From 10.42.0.1 icmp_seq=3 Destination Port Unreachable
```

From Ingress end points,  I am not able to access ingress endpoint from other containers with in the same app  , from containers with in the same project and from  containers from different projects . I am able to access  ingress endpoint only from with in the same container that published the ingress. 

@g-linville Could you please confirm the behavior seen for Ingress endpoints ?

**Comment by g-linville** [comment]
> From Ingress end points, I am not able to access ingress endpoint from other containers with in the same app , from containers with in the same project and from containers from different projects . I am able to access ingress endpoint only from with in the same container that published the ingress.
> 
> @g-linville Could you please confirm the behavior seen for Ingress endpoints ?

@sangee2004 In local clusters, this is how it should behave, yeah. The reason for this is that all of the `*.local.on-acorn.io` domains resolve to `127.0.0.1` (localhost). So it won't work for trying to connect across pods.

In non-local environments I think it should work fine, but I haven't been able to test that yet.

**Comment by sangee2004** [comment]
Tested with `acorn version v0.6.0-113-g83b7b9f3+83b7b9f3`

On Acorn install , Network policy is enabled by default - `networkPolicies: true`
Project isolation for ping, http and ssh access works as expected . 
project1 - App1 - containers - web and ubuntu
project2 - App2 - containers - web and ubuntu and TestContainerApp - containers - testcon

From testcon 
1. Able to curl App2.web's pod ip
2. Able to ssh App2.web's ubuntu ip
3. Able to ping App2.web's pod ip and App1.web's ubuntu ip

4. Not able to curl App1.web's pod ip 
```
root@myapp:/app# curl 10.42.0.91    
curl: (7) Failed to connect to 10.42.0.91 port 80: Connection refused
```
5. Not able to ssh App1.web's ubuntu ip
```
root@myapp:/app# ssh root@10.42.0.90
ssh: connect to host 10.42.0.90 port 22: Connection refused
```
6. Not able to ping App1.web's pod ip and App1.web's ubuntu ip 
```PING 10.42.0.90 (10.42.0.90) 56(84) bytes of data.
From 10.42.0.1 icmp_seq=1 Destination Port Unreachable
From 10.42.0.1 icmp_seq=2 Destination Port Unreachable
From 10.42.0.1 icmp_seq=3 Destination Port Unreachable
```

**Comment by sangee2004** [comment]
When containers have published ports , then we are able to reach (http access) this container cross project using podip. This is as expected.

Having ingress-controller-namespace set - `acorn install --ingress-controller-namespace=kube-system` will result in containers that have published ports , to be not reachable (http access)  cross project using podip.
